File: python-packages/lcyt/lcyt/backend_sender.py
"""Backend relay caption sender — communicates with lcyt-backend instead of YouTube directly."""

import json
import urllib.error
import urllib.request
from typing import Any

from .errors import NetworkError


class BackendCaptionSender:
    """Send live captions via an lcyt-backend relay server.

    Mirrors the ``YoutubeLiveCaptionSender`` API but communicates with an
    ``lcyt-backend`` HTTP server instead of directly with YouTube. This is
    useful when YouTube's API is not reachable from the client (e.g. browser
    environments blocked by CORS, or Python scripts on a restricted network).

    Example:
        >>> sender = BackendCaptionSender(
        ...     backend_url="https://captions.example.com",
        ...     api_key="a1b2c3d4-...",
        ...     stream_key="YOUR_YOUTUBE_KEY",
        ... )
        >>> sender.start()
        >>> sender.send("Hello!")
        >>> sender.send("Relative", time=5000)  # 5 sec since session start
        >>> result = sender.sync()
        >>> status = sender.heartbeat()
        >>> sender.end()
    """

    def __init__(
        self,
        backend_url: str,
        api_key: str,
        stream_key: str,
        domain: str = "http://localhost",
        sequence: int = 0,
        verbose: bool = False,
    ):
        """Initialize the backend relay sender.

        Args:
            backend_url: Base URL of the lcyt-backend server
                         (e.g. ``"https://captions.example.com"``).
            api_key: API key registered in the backend's SQLite database.
            stream_key: YouTube stream key.
            domain: CORS origin that the session will be associated with.
                    Defaults to ``"http://localhost"``.
            sequence: Starting sequence number (overridden by server on start()).
            verbose: Enable verbose logging.
        """
        self._backend_url = backend_url.rstrip("/")
        self._api_key = api_key
        self._stream_key = stream_key
        self._domain = domain
        self._sequence = sequence
        self._verbose = verbose

        self._is_started = False
        self._sync_offset: int = 0
        self._started_at: float = 0.0
        self._token: str | None = None
        self._queue: list[dict] = []

    # ------------------------------------------------------------------
    # Internal fetch helper
    # ------------------------------------------------------------------

    def _fetch(
        self,
        path: str,
        method: str = "GET",
        body: dict | None = None,
        auth: bool = True,
    ) -> dict:
        """Make a JSON request to the backend.

        Args:
            path: Endpoint path (e.g. ``"/live"``).
            method: HTTP method.
            body: Request body (serialised to JSON).
            auth: Attach ``Authorization: Bearer`` header if token is available.

        Returns:
            Parsed JSON response dict.

        Raises:
            NetworkError: On non-2xx response or network failure.
        """
        url = f"{self._backend_url}{path}"
        data = json.dumps(body).encode() if body is not None else None

        headers = {"Content-Type": "application/json"}
        if auth and self._token:
            headers["Authorization"] = f"Bearer {self._token}"

        req = urllib.request.Request(url, data=data, headers=headers, method=method)

        try:
            with urllib.request.urlopen(req) as resp:
                return json.loads(resp.read())
        except urllib.error.HTTPError as exc:
            try:
                error_body = json.loads(exc.read())
                msg = error_body.get("error", f"HTTP {exc.code}")
            except Exception:
                msg = f"HTTP {exc.code}"
            raise NetworkError(msg, exc.code) from exc
        except Exception as exc:
            raise NetworkError(str(exc)) from exc

    # ------------------------------------------------------------------
    # Lifecycle
    # ------------------------------------------------------------------

    def start(self) -> "BackendCaptionSender":
        """Register a session with the backend and obtain a JWT.

        Updates ``sequence``, ``sync_offset``, and ``started_at`` from the
        server response. Idempotent — if a session already exists the server
        returns the existing JWT.

        Returns:
            Self for method chaining.

        Raises:
            NetworkError: If the registration request fails.
        """
        data = self._fetch(
            "/live",
            method="POST",
            body={
                "apiKey": self._api_key,
                "streamKey": self._stream_key,
                "domain": self._domain,
                "sequence": self._sequence,
            },
            auth=False,
        )

        self._token = data["token"]
        self._sequence = data["sequence"]
        self._sync_offset = data["syncOffset"]
        self._started_at = data["startedAt"]
        self._is_started = True

        return self

    def end(self) -> "BackendCaptionSender":
        """Tear down the backend session and clear the stored JWT.

        Returns:
            Self for method chaining.

        Raises:
            NetworkError: If the request fails.
        """
        self._fetch("/live", method="DELETE")
        self._token = None
        self._is_started = False
        return self

    # ------------------------------------------------------------------
    # Caption sending
    # ------------------------------------------------------------------

    def send(
        self,
        text: str,
        timestamp: str | None = None,
        time: int | None = None,
    ) -> dict:
        """Send a single caption.

        Args:
            text: Caption text.
            timestamp: Absolute timestamp string (ISO format). Mutually exclusive
                       with ``time``.
            time: Milliseconds since session start. Resolved server-side using
                  ``startedAt + time + syncOffset``. Mutually exclusive with
                  ``timestamp``.

        Returns:
            Backend response dict: ``{sequence, timestamp, statusCode, serverTimestamp}``.

        Raises:
            NetworkError: If the request fails.
        """
        caption: dict[str, Any] = {"text": text}
        if time is not None:
            caption["time"] = time
        elif timestamp is not None:
            caption["timestamp"] = timestamp

        data = self._fetch("/captions", method="POST", body={"captions": [caption]})
        self._sequence = data.get("sequence", self._sequence)
        return data

    def send_batch(self, captions: list[dict] | None = None) -> dict:
        """Send multiple captions in one request.

        If ``captions`` is None, drains and sends the internal queue built
        with ``construct()``.

        Args:
            captions: List of caption dicts. Each may have ``text``,
                      ``timestamp`` (ISO string), or ``time`` (ms since start).
                      If None, the internal queue is used.

        Returns:
            Backend response dict: ``{sequence, count, statusCode, serverTimestamp}``.

        Raises:
            NetworkError: If the request fails.
        """
        if captions is None:
            items = list(self._queue)
            self._queue.clear()
        else:
            items = captions

        data = self._fetch("/captions", method="POST", body={"captions": items})
        self._sequence = data.get("sequence", self._sequence)
        return data

    # ------------------------------------------------------------------
    # Local queue (construct / send_batch pattern)
    # ------------------------------------------------------------------

    def construct(
        self,
        text: str,
        timestamp: str | None = None,
        time: int | None = None,
    ) -> int:
        """Add a caption to the local queue without sending.

        Call ``send_batch()`` (with no arguments) to flush the queue.

        Args:
            text: Caption text.
            timestamp: Optional absolute timestamp string.
            time: Optional ms-since-session-start offset.

        Returns:
            Current queue length.
        """
        item: dict[str, Any] = {"text": text}
        if time is not None:
            item["time"] = time
        elif timestamp is not None:
            item["timestamp"] = timestamp
        self._queue.append(item)
        return len(self._queue)

    def get_queue(self) -> list[dict]:
        """Return a copy of the current local queue."""
        return list(self._queue)

    def clear_queue(self) -> int:
        """Clear the local queue.

        Returns:
            Number of items cleared.
        """
        count = len(self._queue)
        self._queue.clear()
        return count

    # ------------------------------------------------------------------
    # Sync and heartbeat
    # ------------------------------------------------------------------

    def sync(self) -> dict:
        """Trigger an NTP-style clock sync on the backend.

        Updates local ``sync_offset`` from the response.

        Returns:
            dict: ``{syncOffset, roundTripTime, serverTimestamp, statusCode}``.

        Raises:
            NetworkError: If the request fails.
        """
        data = self._fetch("/sync", method="POST")
        self._sync_offset = data.get("syncOffset", self._sync_offset)
        return data

    def heartbeat(self) -> dict:
        """Check session status on the backend.

        Updates local ``sequence`` and ``sync_offset``.

        Returns:
            dict: ``{sequence, syncOffset}``.

        Raises:
            NetworkError: If the request fails.
        """
        data = self._fetch("/live")
        self._sequence = data.get("sequence", self._sequence)
        self._sync_offset = data.get("syncOffset", self._sync_offset)
        return data

    # ------------------------------------------------------------------
    # Getters / setters
    # ------------------------------------------------------------------

    def get_sequence(self) -> int:
        """Get the current sequence number."""
        return self._sequence

    def set_sequence(self, seq: int) -> "BackendCaptionSender":
        """Set the sequence number manually."""
        self._sequence = seq
        return self

    def get_sync_offset(self) -> int:
        """Get the current sync offset in milliseconds."""
        return self._sync_offset

    def set_sync_offset(self, offset: int) -> "BackendCaptionSender":
        """Set the sync offset manually."""
        self._sync_offset = offset
        return self

    def get_started_at(self) -> float:
        """Get the session start timestamp (Unix epoch seconds from server)."""
        return self._started_at

    @property
    def is_started(self) -> bool:
        """Check if the session is active."""
        return self._is_started

File: packages/lcyt/src/backend-sender.js
import { NetworkError } from './errors.js';

/**
 * Client for sending live captions via an lcyt-backend relay.
 *
 * Mirrors the YoutubeLiveCaptionSender API but communicates with an
 * lcyt-backend HTTP server instead of directly with YouTube. Uses
 * fetch() — works in browsers and Node 18+.
 *
 * @example
 * // Target-array mode (recommended): all targets configured server-side via CC → Targets
 * const sender = new BackendCaptionSender({
 *   backendUrl: 'https://captions.example.com',
 *   apiKey: 'a1b2c3d4-...',
 * });
 * await sender.start({ targets: [{ id: '1', type: 'youtube', streamKey: 'YOUR_KEY' }] });
 *
 * // Legacy single-target mode: pass a streamKey directly
 * const sender = new BackendCaptionSender({
 *   backendUrl: 'https://captions.example.com',
 *   apiKey: 'a1b2c3d4-...',
 *   streamKey: 'YOUR_YOUTUBE_KEY'
 * });
 * await sender.start();
 *
 * await sender.send('Hello!');
 * await sender.sync();
 * await sender.end();
 */
export class BackendCaptionSender {
  /**
   * @param {object} options
   * @param {string} options.backendUrl - Base URL of the lcyt-backend server (e.g. "https://captions.example.com")
   * @param {string} options.apiKey - API key registered in the backend's SQLite database
   * @param {string} [options.authToken] - Project-scoped JWT used for authenticated project requests
   * @param {string} [options.streamKey] - YouTube stream key (optional; superseded by the `targets` array in `start()`)
   * @param {string} [options.domain] - CORS origin. Defaults to location.origin in browsers or 'http://localhost' in Node.
   * @param {number} [options.sequence=0] - Starting sequence number (overridden by backend response on start())
   * @param {boolean} [options.verbose=false] - Enable verbose logging
   */
  constructor({
    backendUrl,
    apiKey,
    authToken,
    streamKey,
    domain,
    sequence = 0,
    verbose = false
  } = {}) {
    this.backendUrl = backendUrl;
    this.apiKey = apiKey;
    this.authToken = authToken || null;
    this.streamKey = streamKey;
    this.domain = domain ||
      (typeof globalThis.location !== 'undefined' ? globalThis.location.origin : 'http://localhost');
    this.sequence = sequence;
    this.verbose = verbose;

    this.isStarted = false;
    this.syncOffset = 0;
    this.startedAt = 0;

    this._token = null;
    this._queue = [];
  }

  // ---------------------------------------------------------------------------
  // Internal fetch helper
  // ---------------------------------------------------------------------------

  /**
   * Make an authenticated JSON request to the backend.
   *
   * @param {string} path - Endpoint path (e.g. '/live')
   * @param {object} [options]
   * @param {string} [options.method='GET']
   * @param {object} [options.body] - Request body (serialised to JSON)
   * @param {boolean} [options.auth=true] - Attach Authorization header if token is available
   * @returns {Promise<object>} Parsed JSON response
   * @throws {NetworkError} On non-2xx response or network failure
   */
  _getAuthToken(auth = true) {
    if (!auth) return null;
    return this.authToken || this._token || null;
  }

  async _fetch(path, { method = 'GET', body, auth = true } = {}) {
    const headers = { 'Content-Type': 'application/json' };
    const token = this._getAuthToken(auth);
    if (token) {
      headers['Authorization'] = 'Bearer ' + token;
    }


    const res = await fetch(`${this.backendUrl}${path}`, {
      method,
      headers,
      body: body ? JSON.stringify(body) : undefined
    });

    const data = await res.json();
    if (!res.ok) {
      throw new NetworkError(data.error || `HTTP ${res.status}`, res.status);
    }
    return data;
  }

  // ---------------------------------------------------------------------------
  // Lifecycle
  // ---------------------------------------------------------------------------

  /**
   * Register a session with the backend and obtain a JWT.
   * Updates sequence, syncOffset, and startedAt from the server response.
   *
   * @param {object} [options]
   * @param {Array} [options.targets] - Optional array of extra caption targets to register with the session.
   *   Each entry: { id, type: 'youtube'|'generic', streamKey?, url?, headers? }
   * @returns {Promise<this>}
   */
  async start({ targets } = {}) {
    const body = {
      apiKey: this.apiKey,
      domain: this.domain,
      sequence: this.sequence
    };
    // streamKey is optional: include it only when provided (legacy single-target mode).
    if (this.streamKey !== null && this.streamKey !== undefined && this.streamKey !== '') {
      body.streamKey = this.streamKey;
    }
    if (Array.isArray(targets) && targets.length > 0) {
      body.targets = targets;
    }
    const data = await this._fetch('/live', {
      method: 'POST',
      body,
      auth: true
    });

    this._token = data.token;
    this.sequence = data.sequence;
    this.syncOffset = data.syncOffset;
    this.startedAt = data.startedAt;
    this.graphicsEnabled = data.graphicsEnabled === true;
    this.isStarted = true;

    return this;
  }

  /**
   * Tear down the backend session and clear the stored JWT.
   * @returns {Promise<this>}
   */
  async end() {
    await this._fetch('/live', { method: 'DELETE' });
    this._token = null;
    this.isStarted = false;
    return this;
  }

  // ---------------------------------------------------------------------------
  // Caption sending
  // ---------------------------------------------------------------------------

  /**
   * Send a single caption.
   *
   * @param {string} text - Caption text
   * @param {string|Date|number|{time: number}} [timestampOrOptions]
   *   - Absolute timestamp (string/Date/epoch ms) — passed as `timestamp`
   *   - `{ time: number }` — milliseconds since session start, resolved server-side
   *   - Omit for auto-generated timestamp
   * @param {object} [extraOpts] - Optional extra fields to merge into the caption object
   *   e.g. { translations: { 'fi-FI': '...' }, captionLang: 'fi-FI', showOriginal: true,
   *          fileFormats: { original: 'vtt', 'fi-FI': 'vtt' } }
   *   `fileFormats` selects the backend caption-file format per language
   *   ('text' | 'youtube' | 'vtt'; 'original' keys the untranslated text)
   * @returns {Promise<{ok: boolean, requestId: string}>} Immediate ack; delivery result arrives via GET /events SSE stream
   */
  async send(text, timestampOrOptions, extraOpts) {
    const caption = { text };

    if (timestampOrOptions !== undefined) {
      if (
        typeof timestampOrOptions === 'object' &&
        timestampOrOptions !== null &&
        'time' in timestampOrOptions
      ) {
        caption.time = timestampOrOptions.time;
      } else {
        caption.timestamp = timestampOrOptions;
      }
    }

    if (extraOpts) {
      if (extraOpts.translations && Object.keys(extraOpts.translations).length > 0)
        caption.translations = extraOpts.translations;
      if (extraOpts.captionLang) caption.captionLang = extraOpts.captionLang;
      if (extraOpts.showOriginal !== undefined) caption.showOriginal = extraOpts.showOriginal;
      if (extraOpts.codes && typeof extraOpts.codes === 'object') caption.codes = extraOpts.codes;
      if (extraOpts.fileFormats && typeof extraOpts.fileFormats === 'object') caption.fileFormats = extraOpts.fileFormats;
    }

    return this._fetch('/captions', {
      method: 'POST',
      body: { captions: [caption] }
    });
  }

  /**
   * Send multiple captions in one request.
   * If no array is provided, drains and sends the internal queue (built with construct()).
   *
   * @param {Array<{text: string, timestamp?: string|Date|number, time?: number}>} [captions]
   * @returns {Promise<{ok: boolean, requestId: string}>} Immediate ack; delivery result arrives via GET /events SSE stream
   */
  async sendBatch(captions) {
    const items = captions !== undefined ? captions : [...this._queue];
    if (captions === undefined) {
      this._queue = [];
    }

    return this._fetch('/captions', {
      method: 'POST',
      body: { captions: items }
    });
  }

  // ---------------------------------------------------------------------------
  // Local queue (construct/sendBatch pattern)
  // ---------------------------------------------------------------------------

  /**
   * Add a caption to the local queue without sending.
   * Use sendBatch() (with no arguments) to flush the queue.
   *
   * When `timestamp` is omitted, the caption is stamped with the current time
   * at queue time (ISO string, no trailing 'Z') so batched captions keep
   * their real spacing instead of collapsing to the flush time.
   *
   * @param {string} text - Caption text
   * @param {string|Date|number|{time: number}} [timestamp] - Optional timestamp,
   *   or `{ time }` (ms since session start, resolved server-side)
   * @param {object} [extraOpts] - Same whitelisted fields as send():
   *   translations, captionLang, showOriginal, codes, fileFormats
   * @returns {number} Current queue length
   */
  construct(text, timestamp, extraOpts) {
    const caption = { text };

    if (timestamp !== undefined && timestamp !== null) {
      if (typeof timestamp === 'object' && !(timestamp instanceof Date) && 'time' in timestamp) {
        caption.time = timestamp.time;
      } else {
        caption.timestamp = timestamp;
      }
    } else {
      caption.timestamp = new Date().toISOString().replace('Z', '');
    }

    if (extraOpts) {
      if (extraOpts.translations && Object.keys(extraOpts.translations).length > 0)
        caption.translations = extraOpts.translations;
      if (extraOpts.captionLang) caption.captionLang = extraOpts.captionLang;
      if (extraOpts.showOriginal !== undefined) caption.showOriginal = extraOpts.showOriginal;
      if (extraOpts.codes && typeof extraOpts.codes === 'object') caption.codes = extraOpts.codes;
      if (extraOpts.fileFormats && typeof extraOpts.fileFormats === 'object') caption.fileFormats = extraOpts.fileFormats;
    }

    this._queue.push(caption);
    return this._queue.length;
  }

  /**
   * Return a copy of the current local queue.
   * @returns {Array<{text: string, timestamp: *}>}
   */
  getQueue() {
    return [...this._queue];
  }

  /**
   * Clear the local queue.
   * @returns {number} Number of items cleared
   */
  clearQueue() {
    const count = this._queue.length;
    this._queue = [];
    return count;
  }

  // ---------------------------------------------------------------------------
  // Sync and heartbeat
  // ---------------------------------------------------------------------------

  /**
   * Trigger an NTP-style clock sync on the backend.
   * Updates local syncOffset from the response.
   * @returns {Promise<object>} { syncOffset, roundTripTime, serverTimestamp, statusCode }
   */
  async sync() {
    const data = await this._fetch('/sync', { method: 'POST' });
    this.syncOffset = data.syncOffset;
    return data;
  }

  /**
   * Check session status on the backend.
   * Updates local sequence and syncOffset.
   * @returns {Promise<{sequence: number, syncOffset: number}>}
   */
  async heartbeat() {
    const data = await this._fetch('/live');
    this.sequence = data.sequence;
    this.syncOffset = data.syncOffset;
    return data;
  }

  /**
   * Update session fields on the backend (e.g. sequence).
   * @param {object} fields - Key/value pairs to PATCH to /live
   * @returns {Promise<object>} Parsed JSON response
   */
  async updateSession(fields) {
    return this._fetch('/live', { method: 'PATCH', body: fields });
  }

  // ---------------------------------------------------------------------------
  // Getters / setters
  // ---------------------------------------------------------------------------

  getSequence() { return this.sequence; }
  setSequence(seq) { this.sequence = seq; return this; }

  getSyncOffset() { return this.syncOffset; }
  setSyncOffset(offset) { this.syncOffset = offset; return this; }

  getStartedAt() { return this.startedAt; }
}
